Document View API reference (#1305)
hyperledger-labs/fabric-smart-client: docs/platform/view/view-api.md

The public contracts are defined in the `platform/view/view` package. The runtime implementation, factory registration, and responder dispatching live in `platform/view/services/view`.

This document focuses on the public API and its runtime semantics. For the internal orchestration model, see the [View service](services/view-service.md). For the platform-level architecture, see the [View SDK architecture](view-sdk.md).

## Overview

At the API level, a view-based FSC application is built from four concepts:

- **View**: A unit of protocol logic implemented by application code.
- **Context**: The execution environment supplied to a running view.
- **Session**: A logical communication channel to another FSC node.
- **Identity**: The local or remote identity bound to a context or session.

The typical lifecycle is:

1. A caller resolves or constructs a view instance.
2. The runtime creates a `view.Context`.
3. The view executes through `Call(context)`.
4. The view uses the context to retrieve services, open sessions, invoke nested views, and manage cleanup.

> 5. In the standard `InitiateView` and responder execution paths, the runtime disposes the context when the protocol branch completes. In explicit `InitiateContext*` flows, the caller owns the context lifecycle and is responsible for deleting it when it is no longer needed.

Reviewer: ```suggestion
5. In the standard `InitiateView` and responder execution paths, the runtime disposes the context when the protocol branch completes.
```

## Packages

Two packages are relevant when writing and exposing views:

- `platform/view/view`: Public contracts for view code.
- `platform/view/services/view`: Runtime implementation, factory registration, and responder registration.

The first package is the programming API. The second package is the runtime that instantiates and dispatches those contracts.

## View

The core abstraction is `view.View`:

```go
type View interface {
    Call(context Context) (interface{}, error)
}
```

> A view is a unit of distributed application logic, A view may:

Reviewer: ```suggestion
A view is a unit of distributed application logic. A view may:
```

```go
type Session interface {
    Info() SessionInfo
    Send(payload []byte) error
    SendWithContext(ctx context.Context, payload []byte) error
    SendError(payload []byte) error
    SendErrorWithContext(ctx context.Context, payload []byte) error
    Receive() <-chan *Message
    Close()
}
```

### Send and Receive

`Send` and `SendWithContext` transmit an application payload to the remote party.

`SendError` and `SendErrorWithContext` transmit an error response. These methods set the message status to `view.ERROR`, allowing the recipient to distinguish protocol failures from normal payloads.

`Receive()` returns a channel of incoming `Message` values. Views typically consume it with a `select` and an explicit timeout rather than waiting indefinitely.

`Close()` releases resources associated with the logical session.

### Message

Incoming messages are represented as:

```go
type Message struct {
    SessionID    string
    ContextID    string
    Caller       string
    FromEndpoint string
    FromPKID     []byte
    Status       int32
    Payload      []byte
    Ctx          context.Context
}
```

The `Status` field uses the constants:

- `view.OK`

> - `view.ERROR`

Reviewer: ```suggestion
- `view.OK` with value `200`
- `view.ERROR` with value `500`
```

This distinction is important when reading logs and runtime documentation. A factory ID such as `create` is not the same thing as the internal identifier of the resulting view type.

## Execution Roles

At the API level, a view usually executes in one of two roles.

### Initiator

An initiator view starts a protocol branch and opens sessions to other parties.

Typical characteristics:

- `Initiator()` returns the initiating view.
- `Session()` is usually `nil`.
- outbound sessions are opened with `GetSession(...)`.

### Responder

A responder view is usually executed because an incoming protocol step arrived from another FSC node. The same responder-style semantics can also be introduced locally through nested execution options such as `AsResponder(session)`.

Typical characteristics:

- `Session()` returns the default responder session.
- the view usually begins by receiving or decoding input from that session;
- the runtime determines the responder through the registration map.

Views can temporarily switch role through nested execution options such as `AsResponder(session)` and `AsInitiator()`.

## Minimal Example

> The `pingpong` integration example is the smallest complete reference for the View API.

Reviewer: ```suggestion
This example shows two view implementations as a reference of the View API adapted from `integration/fsc/pingping`.
```
